deno/lib/types.ts

```diff
@@ -855,6 +856,29 @@ export class ZodString extends ZodType<string, ZodStringDef, string> {
         input.data = input.data.toLowerCase();
       } else if (check.kind === "toUpperCase") {
         input.data = input.data.toUpperCase();
+      } else if (check.kind === "hostname") {
+        const domainNameRegex =
+          /^(?!-)(?!.*--)(?!.*\.\.)(?!.*\.$)[a-zA-Z0-9-]{1,63}(?<!-)(\.[a-zA-Z0-9-]{1,63})*$/;
+        const punycodeRegex = /^xn--[a-zA-Z0-9-]{1,63}$/;
+        const ipv4Regex =
+          /^(25[0-5]|2[0-4][0-9]|[0-1]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[0-1]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[0-1]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[0-1]?[0-9][0-9]?)$/;
+        const ipv6Regex = /^\[[0-9a-fA-F:]+\]$/;
+
+        const isValid =
+          domainNameRegex.test(input.data) ||
+          ipv4Regex.test(input.data) ||
+          ipv6Regex.test(input.data) ||
+          punycodeRegex.test(input.data);
```

Comment:
Also maybe we should allow passing down of an `options` variable that allows us to selectively disable one of these